chore(docs): Update tutorial (#934)
informalsystems/malachite: docs/tutorials/channels.md

itertools = "0.14"
prost = "0.13"
rand = { version = "0.8.5", features = ["std_rng"] }
redb = "2.4.0"
serde = "1.0"
serde_json = "1.0"
sha3 = "0.10"
thiserror = { version = "2.0", default-features = false }
tokio = "1.44.1"
toml = "0.8.19"
tracing = "0.1.41"

[dependencies.malachitebft-app-channel]
version = "0.0.1"
# This adds the `informalsystems-malachitebft-app-channel` as a dependency, but exposes it
# under `malachitebft_app_channel` instead of its full package name.
git = "https://git@github.com/informalsystems/malachite.git"
package = "informalsystems-malachitebft-app-channel"

[dependencies.malachitebft-proto]
version = "0.0.1"
git = "https://git@github.com/informalsystems/malachite.git"
package = "informalsystems-malachitebft-proto"

[dependencies.malachitebft-test]
version = "0.0.1"
git = "https://git@github.com/informalsystems/malachite.git"
package = "informalsystems-malachitebft-test"

[dependencies.malachitebft-test-cli]
version = "0.0.1"
git = "https://git@github.com/informalsystems/malachite.git"
package = "informalsystems-malachitebft-test-cli"
```

### Application state

Before handling consensus messages, let's start by preparing the application state.

> ```

Reviewer: ```suggestion
```rust
```

// src/main.rs

mod state;
mod store;
mod streaming;
```

```rust
// src/state.rs

pub struct State {
    ctx: TestContext,
    signing_provider: Ed25519Provider,
    genesis: Genesis,
    address: Address,
    vote_extensions: HashMap<Height, VoteExtensions<TestContext>>,
    streams_map: PartStreamsMap,
    rng: StdRng,

    pub store: Store,
    pub current_height: Height,
    pub current_round: Round,
    pub current_proposer: Option<Address>,
    pub peers: HashSet<PeerId>,
}
```

> We will use two other crates called `store` and `streaming`. The `store` crate provide a database called `Store` which uses [redb](https://docs.rs/redb/latest/redb/) under the hood and offer the following interface:

Reviewer: ```suggestion
We will use two other modules called `store` and `streaming`. The `store` crate provide a database called `Store` which uses [redb](https://docs.rs/redb/latest/redb/) under the hood and offer the following interface:
```

    /// Store a decided value
    pub async fn store_decided_value(
        &self,
        certificate: &CommitCertificate<TestContext>,
        value: Value,
    ) -> Result<(), StoreError>

    /// Store an undecided proposal
    pub async fn store_undecided_proposal(
        &self,
        value: ProposedValue<TestContext>,
    ) -> Result<(), StoreError>

    /// Get the undecided proposal at the given height and round
    pub async fn get_undecided_proposal(
        &self,
        height: Height,
        round: Round,
    ) -> Result<Option<ProposedValue<TestContext>>, StoreError>

    /// Prune the store, removing all decided values below the given height
    pub async fn prune(&self, retain_height: Height) -> Result<Vec<Height>, StoreError>

    /// Remove undecided proposals matching the given value id
    pub async fn remove_undecided_proposals_by_value_id(
        &self,
        value_id: ValueId,
    ) -> Result<(), StoreError>

    /// Get the undecided proposal matching the given value id
    pub async fn get_undecided_proposal_by_value_id(
        &self,
        value_id: ValueId,
    ) -> Result<Option<ProposedValue<TestContext>>, StoreError>
```

> Note that the implementation of this store is up to the application developer. One could have used a simple in-memory store.

Reviewer: The store must be persistent, otherwise the application cannot be restarted in some cases.

```suggestion
Note that the implementation of this store is up to the application developer, who may choose a different underlying database such as RocksDB, LevelDB, or [Fjall](https://github.com/fjall-rs/fjall).
```

> The `streaming` crate provides a `PartStreamsMap` data structure. This is used to keep track of the proposal parts that are being streamed over the network. It is also used to re-assemble the full proposal once all parts have been received. It provides the following interface:

Reviewer: ```suggestion
The `streaming` module provides a `PartStreamsMap` data structure. This is used to keep track of the proposal parts that are being streamed over the network. It is also used to re-assemble the full proposal once all parts have been received. It provides the following interface:
```

```rust
    /// Initialize the data structure
    pub fn new() -> Self

    /// Insert a proposal part into the map, returning the full proposal if all parts have been received
    pub fn insert(
        &mut self,
        peer_id: PeerId,
        msg: StreamMessage<ProposalPart>,
    ) -> Option<ProposalParts>
```

> Please refer to the respective crate files for the full implementation of the [`store`](/code/examples/channel/src/store.rs) and [`streaming`](/code/examples/channel/src/streaming.rs) crates.

Reviewer: ```suggestion
Please refer to [`store`](/code/examples/channel/src/store.rs) and [`streaming`](/code/examples/channel/src/streaming.rs) modules for their full implementation.
```

> Now, we can go through the implementation of the application state. Let's start with helper methods that will be used by the state implementation. Note that the way a proposal is split here is specific to our case (where the value is an integer). In a real application, the value is likely to be more complex and the splitting logic would be different.

Reviewer: ```suggestion
Now, we can go through the implementation of the application state. Let's start with helper methods that will be used by the state implementation. Note that the way a proposal is split here is specific to our case (where the value is a natural number and we split it by factoring it into its prime factors). In a real application, the value is likely to be more complex and the splitting logic would be different.
```

        pol_round: Round,
    ) -> Vec<ProposalPart> {
        let mut hasher = sha3::Keccak256::new();
        let mut parts = Vec::new();

        // Init
        // Include metadata about the proposal
        {
            parts.push(ProposalPart::Init(ProposalInit::new(
                value.height,
                value.round,
                pol_round,
                self.address,
            )));

            hasher.update(value.height.as_u64().to_be_bytes().as_slice());
            hasher.update(value.round.as_i64().to_be_bytes().as_slice());
        }

        // Data
        // Include each prime factor of the value as a separate proposal part
        {
            for factor in factor_value(value.value) {
                parts.push(ProposalPart::Data(ProposalData::new(factor)));

                hasher.update(factor.to_be_bytes().as_slice());
            }
        }

        // Fin
        // Sign the hash of the proposal parts
        {
            let hash = hasher.finalize().to_vec();
            let signature = self.signing_provider.sign(&hash);
            parts.push(ProposalPart::Fin(ProposalFin::new(signature)));
        }

        parts
    }

>     /// Creates a stream message containing a proposal part.

Reviewer 2: ```suggestion
    /// Creates a stream of messages containing a proposal parts.
```

                    certificate: decided_value.certificate,
                    value_bytes: ProtobufCodec.encode(&decided_value.value).unwrap(),
                });

                if reply.send(raw_decided_value).is_err() {
                    error!("Failed to send GetDecidedValue reply");
                }
            }
```

The `malachite-test-cli` crate uses the `clap` library to parse the command-line and provide input parameters like
`--home` as well as commands like `init`, `start` or `testnet`.
In order to figure out if we can help a peer that is lagging behind,
the engine may ask us for the height of the earliest available value in our store.

```rust
            AppMsg::GetHistoryMinHeight { reply } => {
                let min_height = state.get_earliest_height().await;

These work similarly to CometBFT: `init` creates initial configuration in `$HOME/.malachite`, `testnet` does the same but
it creates multiple configurations (make sure `--nodes` is set) for running a local testnet, and `start` runs the node.
                if reply.send(min_height).is_err() {
                    error!("Failed to send GetHistoryMinHeight reply");
                }
            }
```

> ### Logging

Reviewer 3: Dedicated logging section might still have been useful, no?

Author: Right, I will add it to the command-line section, where it is initialized.

                    error!("Failed to send ExtendVote reply");
                }
            }
```

The application is also responsible to verify a given vote extension. In our case, we simply return `OK(())`.

```rust
            AppMsg::VerifyVoteExtension {
                height: _,
                round: _,
                value_id: _,
                extension: _,
                reply,
            } => {
                if reply.send(Ok(())).is_err() {
                    error!("Failed to send VerifyVoteExtension reply");
                }
            }
```

In some cases, e.g. to verify the signature of a vote received at a higher height
than the one we are at (e.g. because we are lagging behind a little bit),
the engine may ask us for the validator set at that height.

In our case, our validator set stays constant between heights so we can
send back the validator set found in our genesis state.

```rust
            AppMsg::GetValidatorSet { height: _, reply } => {
                if reply.send(state.get_validator_set().clone()).is_err() {
                    error!("Failed to send GetValidatorSet reply");
                }
            }
```

> As just mentioned, tt may happen that our node is lagging behind its peers. In that case,

Reviewer 2: ```suggestion
As just mentioned, it may happen that our node is lagging behind its peers. In that case,
```
